### PR TITLE
[DO NOT MERGE] Base Implementation of Transcriptomics Project

### PR DESCRIPTION
**Pull request by nbollis**
Started implementation of the transcriptomics project in MzLib. 

The scope of this PR is to create the base classes to expand upon in subsequent PR's and the ability to fragment the oligos. 

This mzLib was packed and added to the current branch of MetaMorpheus, only two things were needed to make the code functional and all tests pass. 
- Change all instances of Product to IProduct
- Change references from Proteomics.FragmentationTerminus  and Proteomics.ProductType to MassSpectrometry.FragmentationTerminus and MassSpectrometry.ProductType

**Changes to previous codebase:**
Product Type
 > Moved from Proteomics to MassSpectrometery
 > Added a bunch of new product ion types

FragmentationTerminus
 > Moved from Proteomics to MassSpectrometery
 > Added FivePrime and ThreePrime

IProduct
 > Created an interface for product (fragment) ions in MassSpectrometery
 > Enables Protein product ions and RnaProduct Ions to be treated equivalently in MetaMorpheus code
 > Changes many references from Product to IProduct
 
Proteomics.Product
> Inhertis from IProduct Interface
> Changed from struct to class so it is always passed by reference instead of copying values onto the stack each time it is references. 

**New Code:**
Nucleotide 
> Represents a single nucleotide in an oligo
> Has default implementations for all 4 common DNA and RNA bases
> Allows addition of custom bases

NucleicAcid
> Base class which RNA and eventually DNA inherit from
> Contains sequence information in the form of an array of nucleotides and a simple base sequence
> Fragmentation uses identical syntax to PeptideWithSetMods.Fragment()

RNAFragment
> Inherits from IProduct interface
> Can be used in MetaMorpheus code interchangably with Proteomics.Product
> Contains identical property names to the Proteomics.Product class

DissociationTypeCollection
> Implements almost every function in the Proteomics.DissociationTypeCollection including
> -  Get Product Types based upon DissociationType (only CID and HCD are implemented)
> - Get Mass shift based upon product type
> - Get the terminus type based upon the fragment type
> - Get Product Type based upon DissociationType and a specific Terminus (CID and HCD)

### CODE REVIEW

File: mzLib/MassSpectrometry/BaseClasses/FragmentationTerminus.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace MassSpectrometry
+{
+    public enum FragmentationTerminus
+    {
+        Both, //N- and C-terminus
+        N, //N-terminus only
+        C, //C-terminus only
+        None, //used for internal fragments, could be used for top down intact mass?
+        FivePrime, // 5' for NucleicAcids
```
Comment (trishorts) [inline mzLib/MassSpectrometry/BaseClasses/FragmentationTerminus.cs:15]: Do we need a "both" for nucleic acids?

File: mzLib/MassSpectrometry/BaseClasses/IProduct.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        public double NeutralMass { get; }
+        public ProductType ProductType { get; }
+        public double NeutralLoss { get; }
+        public FragmentationTerminus Terminus { get; }
+        public int FragmentNumber { get; }
+        public int ResiduePosition { get; } // added for interface compatibility 
+        public int AminoAcidPosition => ResiduePosition;
+        public ProductType? SecondaryProductType { get; } //used for internal fragment ions
+        public int SecondaryFragmentNumber { get; } //used for internal fragment ions
```
Comment (trishorts) [inline mzLib/MassSpectrometry/BaseClasses/IProduct.cs:20]: should this be nullable? if there is no secondary product type

File: mzLib/MassSpectrometry/BaseClasses/ProductType.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        //    //y°	        y-H2O
+        //    //z             [C]+[M]-NH2
+
+
+
+        a,
+        aStar,
+        aDegree,
+        aWaterLoss,
+        aBase,
```
Comment (trishorts) [inline mzLib/MassSpectrometry/BaseClasses/ProductType.cs:35]: does Base mean nucleotide? and is it the same for rna and dna?

File: mzLib/Proteomics/Fragmentation/DissociationTypeCollection.cs [hunk 4]
```diff
@@ -102,6 +102,25 @@ public static List<ProductType> GetWaterAndAmmoniaLossProductTypesFromDissociati
             { ProductType.D, null},// diagnostic ions are not shifted but added sumarily
             { ProductType.Ycore, null},// neutral Molecular product can be used with neutral loss as fragment
             { ProductType.Y, null},// diagnostic ions are not shifted but added sumarily
+
+            // Rna Specific types are not shifted through this method, but added to not break the code.
```
Comment (trishorts) [inline mzLib/Proteomics/Fragmentation/DissociationTypeCollection.cs:106]: why change the standard protocol?

File: mzLib/Proteomics/Fragmentation/Product.cs [hunk 5]
```diff
@@ -32,67 +41,52 @@ public struct Product
             SecondaryFragmentNumber = secondaryFragmentNumber;
         }
 
-        public string Annotation
```
Comment (trishorts) [inline mzLib/Proteomics/Fragmentation/Product.cs:35]: can i assume this was unused?

File: mzLib/Transcriptomics/Classes/NucleicAcid.cs [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+    /// </summary>
+    public abstract class NucleicAcid : INucleicAcid, IEquatable<NucleicAcid>, IHasMass
+    {
+
+        #region Static Properties
+
+        /// <summary>
+        /// The default chemical formula of the five prime (hydroxyl group)
+        /// </summary>
+        public static readonly ChemicalFormula DefaultFivePrimeTerminus = ChemicalFormula.ParseFormula("O-3P-1");
```
Comment (trishorts) [inline mzLib/Transcriptomics/Classes/NucleicAcid.cs:26]: is it clear what "Default" means?

File: mzLib/Transcriptomics/Classes/NucleicAcid.cs [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+        private IHasChemicalFormula _5PrimeTerminus;
+
+        /// <summary>
+        /// The 3-Prime chemical formula cap. This is different from the 3-prime terminus modification.
+        /// </summary>
+        private IHasChemicalFormula _3PrimeTerminus;
+
+        /// <summary>
+        /// All of the modifications indexed by position from 5- to 3-prime termini. This array is 2 bigger than the nucleic acid array
+        /// as index 0 and Count - 1 represent the 5- and 3-prime terminus, respectively
```
Comment (trishorts) [inline mzLib/Transcriptomics/Classes/NucleicAcid.cs:68]: ? where count = length + 2?

Comment (trishorts) [inline mzLib/Transcriptomics/Classes/NucleicAcid.cs:68]: we also have "oneIsNterminus" for peptides (numbering). is this a thing we need for nucleic acids?

File: mzLib/Transcriptomics/Classes/NucleicAcid.cs [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+
+        /// <summary>
+        /// Gets the base nucleic acid sequence
+        /// </summary>
+        public string BaseSequence
+        {
+            get
+            {
+                // Generate the sequence if the stored version is null or empty
+                if (string.IsNullOrEmpty(_sequence))
```
Comment (trishorts) [inline mzLib/Transcriptomics/Classes/NucleicAcid.cs:150]: do we need a check to make sure that the sequence that gets passed is actually composed of acceptable characters? I ran into this problem elsewhere.

**Review by trishorts — COMMENTED** [review]
great start

File: mzLib/MassSpectrometry/BaseClasses/FragmentationTerminus.cs (hunk 1 above)
Comment (nbollis) [reply mzLib/MassSpectrometry/BaseClasses/FragmentationTerminus.cs:15]: Both was already in this enum from proteins. Unsure if it will be used/need for RNA

File: mzLib/MassSpectrometry/BaseClasses/IProduct.cs (hunk 2 above)
Comment (nbollis) [reply mzLib/MassSpectrometry/BaseClasses/IProduct.cs:20]: In the original implementation, this was not nullable, with the default set to zero. This behavior has been maintained as to not throw a bunch of null check warnings in MM when updated to this mzLib

File: mzLib/Proteomics/Fragmentation/DissociationTypeCollection.cs (hunk 4 above)
Comment (nbollis) [reply mzLib/Proteomics/Fragmentation/DissociationTypeCollection.cs:106]: Because both rna and proteins have a a,b,c,x,y,z ions which differ in their neutral losses. One central location for neutral losses would not be possible without specifying which type of bio-polymer you are interested in. Implementing the method in this fashion would require all calls in MetaMorpheus to be altered

File: mzLib/Proteomics/Fragmentation/Product.cs (hunk 5 above)
Comment (nbollis) [reply mzLib/Proteomics/Fragmentation/Product.cs:35]: This is used quite often, but the implementation for how to parse the annotation from the object data has been moved to the interface. The new getter for Annotation (line 23) calls the interfaces implementation of this method. The same is true for the RNAProduct class

File: mzLib/Transcriptomics/Classes/NucleicAcid.cs (hunk 6 above)
Comment (nbollis) [reply mzLib/Transcriptomics/Classes/NucleicAcid.cs:26]: Added some comments for clarity, please let me know if that answers your question

File: mzLib/Transcriptomics/Classes/NucleicAcid.cs (hunk 7 above)
Comment (nbollis) [reply mzLib/Transcriptomics/Classes/NucleicAcid.cs:68]: Ignore for now. This was from the CSMSL implementation and is currently unused. It will be changed to be more in line with out standard method when the modification PR comes around

File: mzLib/Proteomics/Fragmentation/DissociationTypeCollection.cs (hunk 4 above)
Comment (nbollis) [reply mzLib/Proteomics/Fragmentation/DissociationTypeCollection.cs:106]: Comments were added

File: mzLib/Proteomics/Fragmentation/MatchedFragmentIon.cs [hunk 9]
```diff
@@ -1,26 +1,45 @@
...
             NeutralTheoreticalProduct = neutralTheoreticalProduct;
             Mz = experMz;
             Intensity = experIntensity;
             Charge = charge;
         }
 
+        /// <summary>
+        /// Constructs a new MatchedFragmentIon given information about a theoretical and an experimental fragment mass spectral peak
+        /// </summary>
+        public MatchedFragmentIon(ref IProduct neutralTheoreticalProduct, double experMz, double experIntensity, int charge)
```
Comment (Alexander-Sol) [inline mzLib/Proteomics/Fragmentation/MatchedFragmentIon.cs:29]: I believe that the ref keyword is only applicable for reference types if you need to replace it entirely. e.g., the matched fragment ion constructor was creating a new IProduct, assigning it to neutralTheoreticalProduct, and you wanted that new assignment to be visible outside of the local type.

tldr; lose the ref

File: mzLib/Proteomics/Fragmentation/MatchedFragmentIon.cs [hunk 10]
```diff
@@ -1,26 +1,45 @@
...
+
+        /// <summary>
+        /// Constructs a new MatchedFragmentIon given information about a theoretical and an experimental fragment mass spectral peak
+        /// </summary>
+        public MatchedFragmentIon(ref Product neutralTheoreticalProduct, double experMz, double experIntensity,
+            int charge)
+            : this(neutralTheoreticalProduct, experMz, experIntensity, charge)
+        {
+        }
+
```
Comment (Alexander-Sol) [inline mzLib/Proteomics/Fragmentation/MatchedFragmentIon.cs:42]: lose the ref

File: mzLib/Proteomics/Fragmentation/Product.cs [hunk 11]
```diff
@@ -32,67 +41,52 @@ public struct Product
...
+                   SecondaryFragmentNumber == other.SecondaryFragmentNumber &&
+                   MonoisotopicMass.Equals(other.MonoisotopicMass);
+        }
 
-            return this.ProductType == other.ProductType
-                && this.NeutralMass == other.NeutralMass
-                && this.FragmentNumber == other.FragmentNumber
-                && this.NeutralLoss == other.NeutralLoss
-                && this.SecondaryFragmentNumber == other.SecondaryFragmentNumber
-                && this.SecondaryProductType == other.SecondaryProductType;
```
Comment (Alexander-Sol) [inline mzLib/Proteomics/Fragmentation/Product.cs:90]: I really like the previous format. One bool evaluation per line and each line starts with the && operator

File: mzLib/Proteomics/Fragmentation/Product.cs [hunk 12]
```diff
@@ -32,67 +41,52 @@ public struct Product
...
             }
         }
 
-        public override bool Equals(object obj)
+        public bool Equals(IProduct other)
         {
-            Product other = (Product)obj;
+            return NeutralMass.Equals(other.NeutralMass) && ProductType == other.ProductType &&
+                   NeutralLoss.Equals(other.NeutralLoss) && Terminus == other.Terminus &&
+                   FragmentNumber == other.FragmentNumber && Math.Abs(ResiduePosition - other.ResiduePosition) < 0.0001 &&
```
Comment (Alexander-Sol) [inline mzLib/Proteomics/Fragmentation/Product.cs:66]: ResiduePosition is an int. you can just do ResiduePosition == other.ResiduePosition

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs [hunk 13]
```diff
@@ -217,7 +217,7 @@ public string SequenceWithChemicalFormulas
         /// Generates theoretical fragments for given dissociation type for this peptide. 
         /// The "products" parameter is filled with these fragments.
         /// </summary>
-        public void Fragment(DissociationType dissociationType, FragmentationTerminus fragmentationTerminus, List<Product> products)
+        public void Fragment(DissociationType dissociationType, FragmentationTerminus fragmentationTerminus, List<IProduct> products)
         {
             // This code is specifically written to be memory- and CPU -efficient because it is 
             // called millions of times for a typical search (i.e., at least once per peptide).
```
Comment (Alexander-Sol) [inline mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:223]: This comment has me a little worried. Does changing product from struct to class increase or decrease efficiency?

File: mzLib/Transcriptomics/Fragmentation/RnaProduct.cs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+
+        public bool Equals(IProduct other)
+        {
+            if (other is null) return false;
+            return NeutralMass.Equals(other.NeutralMass) && ProductType == other.ProductType &&
+                   NeutralLoss.Equals(other.NeutralLoss) && Terminus == other.Terminus &&
+                   FragmentNumber == other.FragmentNumber && ResiduePosition == other.ResiduePosition &&
+                   SecondaryProductType == other.SecondaryProductType &&
+                   SecondaryFragmentNumber == other.SecondaryFragmentNumber;
```
Comment (Alexander-Sol) [inline mzLib/Transcriptomics/Fragmentation/RnaProduct.cs:65]: one check per line pretty plz w/ sugar on top

**Review by Alexander-Sol — CHANGES_REQUESTED** [review]

File: mzLib/Transcriptomics/Classes/Nucleotide.cs [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        #endregion
+
+        #region Constructors
+
+        internal Nucleotide(string name, char oneLetterAbbreviation, string threeLetterAbbreviation, string baseChemicalFormula)
+            : this(name, oneLetterAbbreviation, threeLetterAbbreviation, ChemicalFormula.ParseFormula(baseChemicalFormula))
+        {
+        }
+
+        public Nucleotide(string name, char oneLetterAbbreviation, string threeLetterAbbreviation,
```
Comment (Alexander-Sol) [inline mzLib/Transcriptomics/Classes/Nucleotide.cs:105]: Can you add comments making it clear that this constructor can throw an Exception if an invalid sequence is passed in?

File: mzLib/Proteomics/Fragmentation/Product.cs (hunk 5 above)
Comment (trishorts) [reply mzLib/Proteomics/Fragmentation/Product.cs:35]: will you show me this some time?

File: mzLib/Transcriptomics/Classes/NucleicAcid.cs (hunk 7 above)
Comment (trishorts) [reply mzLib/Transcriptomics/Classes/NucleicAcid.cs:78]: we should discuss number (we have the "one is n terminus") business for peptides

**Review by trishorts — APPROVED** [review]

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs (hunk 13 above)
Comment (nbollis) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:223]: Added comment to ensure the date of change is noted, definitely something to consider moving forward

File: mzLib/Proteomics/Fragmentation/Product.cs (hunk 12 above)
Comment (nbollis) [reply mzLib/Proteomics/Fragmentation/Product.cs:66]: fixed

File: mzLib/Transcriptomics/Fragmentation/RnaProduct.cs (hunk 14 above)
Comment (nbollis) [reply mzLib/Transcriptomics/Fragmentation/RnaProduct.cs:65]: fixed

**Review by trishorts — APPROVED** [review]

**Review by elaboy — APPROVED** [review]

**Review by Alexander-Sol — DISMISSED** [review]
